satellite: graceful evacuation when using ClusterAPI (#855)
piraeusdatastore/piraeus-operator: docs/explanation/deletion-policies.md

LINSTOR will keep the offline Satellite indefinitely. Resources assigned to the Satellite will still appear in the list
of LINSTOR resources and LINSTOR will keep necessary information such as device paths in its database to restore the
Satellite when it is recreated.

This is the default policy, as it is the least disruptive policies in Clusters that have a fixed set of Nodes, or
clusters where Nodes are only added but not removed.

### `Evacuate` Policy

Using the `Evacuate` deletion policy, Piraeus Datastore will first ensure the configured number of replicas for all
volumes remains available before removing the Satellite. The Satellite is only removed after all local replicas and
snapshots are removed.

In order to keep potential fail-over times short, Piraeus Datastore performs the following steps during evacuation:

1. Piraeus Datastore searches for PersistentVolumes (PVs) that:
    * Are attached on the Node to be evacuated and marks them with `wait-for-reattach.evacuation.piraeus.io/<node-name>`
      annotation. It will later use this annotation to wait for the PV to be reattached.

>     * In addition, if the PVS are also either using the `allowRemoteVolumeAcesss: "false"` policy or are otherwise bound

Reviewer: ```suggestion
    * In addition, if the PVs are also either using the `allowRemoteVolumeAcesss: "false"` policy or are otherwise bound
```

      to the Node to be evacuated in their `.spec.nodeAffinity`, they are annotated with
      `override.piraeus.io/<nodename>`. This causes the LINSTOR Affinity Controller to temporarily mark the PV as
      accessible from any Node.

> 2. Piraeus Datastore will wait for any other `LinstorSatellite` resource to become available. This ensures that during

Reviewer: ```suggestion
2. Piraeus Datastore will wait for all other `LinstorSatellite` resources to become available. This ensures that during
```

Right?

   rolling upgrades of the infrastructure, it waits until a replacement node for the node to be evacuated is available.
3. If using ClusterAPI, the `pre-drain.delete.hook.machine.cluster.x-k8s.io/linstor-prepare-for-drain` annotation on the
   Machine is removed. This will cause ClusterAPI to start draining the node.
4. Piraeus Datastore will wait for all PVs that were marked in Step 1 to be reattached on new nodes. This ensures that
   LINSTOR can choose the optimal replica placement for the next step.
5. Piraeus Datastore will signal LINSTOR to evacuate the node. LINSTOR will create replacement resources for all diskful
   and diskless resources on the node. Volumes that are currently in use on a diskless will get a local replica, if that
   is possible based on the available storage pools.
6. Piraeus Datastore will wait for LINSTOR to complete the evacuation. The evacuated Node no longer has any resources

>    or snapshot.

Reviewer: ```suggestion
   or snapshots.
```

7. All temporary annotations from PVs are removed. In particular, any temporary override that allowed a volume to be
   attached on a node violating the usual access policies is removed.
8. If using ClusterAPI, the `pre-terminate.delete.hook.machine.cluster.x-k8s.io/linstor-wait-for-complete-evacuation`
   annotation on the Machine is removed, letting ClusterAPI proceed with shutting down the node.
9. The Satellite is removed from LINSTOR.

### `Delete` Policy

When using the `Delete` policy, Piraeus Datastore will remove the Satellite from the LINSTOR Cluster in addition to
removing the associated Kubernetes resources. The Satellite will no longer appear in the list of LINSTOR nodes and
any resources on it are removed.

!!! warning "There are no checks preventing deletion of the last replica of a volume"

    Using this policy means that there are no checks the prevent the deletion of the last replica of a volume. While

>     LINSTOR automatically tries to maintain the requested number of replicas, does so only periodically.

Reviewer: ```suggestion
    LINSTOR automatically tries to maintain the requested number of replicas, it does so only periodically.
```
